Add Swagger 2.0 documentation for Prysm API endpoints

willianpaixao opened the pull request:
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This PR adds Swagger 2.0 documentation for all Prysm API endpoints (14 routes total). Custom Prysm endpoints (/prysm/node/*, /prysm/v1/validators/*, /prysm/v1/beacon/*) were previously undocumented, making them difficult for developers to discover and use.

This implements annotation-based documentation using [swaggo/swag](https://github.com/swaggo/swag), generating machine-readable API specification files (swagger.yaml, swagger.json) that can be used for client SDK generation, API exploration, and interactive documentation.

### Technical Details

**Swagger Generation:**
```bash
swag init --parseDependency --parseInternal --generalInfo doc.go \
  --dir ./beacon-chain/rpc/prysm,./api/server/structs,./network/httputil,./consensus-types/primitives \
  --output ./docs
```

**Implementation Notes:**
- Annotations use simple package names
- Dual `@Router` annotations for GetValidatorCount (accessible at both `/prysm/v1` and `/eth/v1` paths)
- No handler logic changes, only documentation comments added
- Generated files can be used with Swagger UI or for client SDK generation

**Which issues(s) does this PR fix?**

Fixes #12794

### Changelog Entry

A changelog fragment file will be added for this user-facing feature addition per CONTRIBUTING.md requirements.

### Future Enhancements

This PR provides the foundation for:
- Serving Swagger UI at `/swagger/index.html` (requires adding http-swagger dependency)
- Upgrading to OAS 3.0 when swag adds support or via conversion tools
- Generating client SDKs in multiple languages using swagger-codegen
- Publishing documentation to GitHub Pages
- CI/CD integration to auto-generate docs on every build

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

willianpaixao committed:
Add Swagger 2.0 documentation for Prysm API endpoints

This commit implements API documentation for all custom
Prysm endpoints using swaggo/swag, addressing GitHub issue #12794.

Changes:
- Added Swagger annotations endpoint handlers
- Generated swagger.yaml, swagger.json, and docs.go in /docs directory

Swagger generation command:
  swag init --parseDependency --parseInternal --generalInfo doc.go \
    --dir ./beacon-chain/rpc/prysm,./api/server/structs,./network/httputil,./consensus-types/primitives \
    --output ./docs

Technical notes:
- All annotations use simple package names
- No handler logic changes, only doc comments added

Closes: #12794

Signed-off-by: Willian Paixao <willian@ufpa.br>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/OffchainLabs/prysm?pullRequest=16080) <br/>All committers have signed the CLA.

willianpaixao commented:
To run swagger-ui with Docker, simply run the following command from the project root folder:

```bash
$ docker run --rm --publish 8080:8080 --env SWAGGER_JSON=/docs/swagger.json --volume ./docs:/docs docker.swagger.io/swaggerapi/swagger-ui
```

Follow some screenshots:

<img width="1430" height="837" alt="Screenshot_20251129_114354" src="https://github.com/user-attachments/assets/b65a9ea8-dec0-4948-95da-d4178fced88b" />
<img width="1432" height="469" alt="Screenshot_20251129_114256" src="https://github.com/user-attachments/assets/ed4dd1f5-63d1-4dae-a5e3-5a9d16110d2f" />
<img width="1438" height="779" alt="Screenshot_20251129_114233" src="https://github.com/user-attachments/assets/3fc98098-b80b-4494-9031-35ba1851fe51" />

willianpaixao committed:
chiore: cleaning up go.mod

Signed-off-by: Willian Paixao <willian@ufpa.br>

willianpaixao commented on go.mod:
```diff
@@ -143,6 +145,21 @@ require (
+	github.com/go-openapi/swag/netutils v0.25.4 // indirect
+	github.com/go-openapi/swag/stringutils v0.25.4 // indirect
+	github.com/go-openapi/swag/typeutils v0.25.4 // indirect
+	github.com/go-openapi/swag/yamlutils v0.25.4 // indirect
```
You have a point, fixed.

willianpaixao commented on go.mod:
```diff
@@ -278,7 +297,7 @@ require (
 	sigs.k8s.io/json v0.0.0-20221116044647-bc3834ca7abd // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.4.1 // indirect
-	sigs.k8s.io/yaml v1.3.0 // indirect
+	sigs.k8s.io/yaml v1.6.0 // indirect
```
Fixed.

james-prysm commented:
Thanks for this but we're not quite sure if this is the right solution for us

We have 2 issues with this, we've discussed a lot about this internally

1. external dependency
2. overhead of adding and keeping up with these annoations 

i wonder if either we can somehow have some script that auto updates this meta information + not have to include the swagger dependency in prysm but some kind of viewer after the json generates would be better. we are really against adding dependencies if not needed for core features for security reasons.

willianpaixao commented:
Thanks for the feedback @james-prysm, these are super valid points! I've addressed both concerns:

  1. Dependency Concern - Resolved

  Removed github.com/swaggo/swag from go.mod entirely. The swagger documentation now uses a CLI-only approach:
  - Deleted docs/docs.go (the file that imported the swag library)
  - Kept swagger.json and swagger.yaml as static files
  - swag is only needed as a CLI tool during development, not as a runtime dependency

  To regenerate the docs:

```bash
  go install github.com/swaggo/swag/cmd/swag@latest
  swag init -g beacon-chain/rpc/prysm/doc.go -o docs/ --parseDependency --parseInternal
```

  2. Maintenance Overhead

  I did a quick _git-blame_ to the handler files for all endpoints. Result: zero changes in the last few months
   that would require annotation updates. These endpoints seemed to be a stable, mature API surface. Estimated maintenance burden is 2-3 annotation updates per year, and changes are typically additive (new endpoints) rather than modifications.

   Added _docs/README.md_ documenting the CLI-based generation workflow.

james-prysm commented:
Thanks currently focused on fusaka items will bring this up to team

willianpaixao commented:
Hey @james-prysm, any updates?

james-prysm commented:
@willianpaixao sorry for late response most of the team was out for holiday, we're not ready to include this just yet, we are still working through some ongoing issues on mainnet nodes as well as any preparation for the upcoming bpo changes. apologies but it may be some time before circling back here.

willianpaixao commented:
@james-prysm I totally understand. Happy new year!
Is that okay if I ping you and the team in a month or two? Not to rush anything, is that I know those low-profile PRs tend to get forgotten.

james-prysm commented:
@willianpaixao sure would appreciate the reminder on that

willianpaixao commented:
Follow some screenshots of the implemented feature.

<img width="1443" height="1927" alt="Screenshot_20260112_012250" src="https://github.com/user-attachments/assets/0f363f07-f927-4c41-b90a-93330bef483c" />

<img width="1439" height="1987" alt="Screenshot_20260112_012521" src="https://github.com/user-attachments/assets/35faa55d-f491-4020-b193-081846f77645" />

<img width="1431" height="1585" alt="Screenshot_20260112_012031" src="https://github.com/user-attachments/assets/0b23fe56-f6c8-49c8-af74-4c0e6b28c56d" />

willianpaixao commented:
Hey @james-prysm maybe we can take a look at this now?

willianpaixao commented:
🦥
